### PR TITLE
XDELEX Command

### PR DESCRIPTION
**Pull request by nickiaq**
This PR adds a stream command `XDELEX` that deletes one or more stream messages with explicit control over how consumer groups' PEL references are handled.  While implementing `XACKDEL`, it made sense to also add this as a complimentary command.

For more background on the problem space, see #2903. Similar to `XACKDEL`, this is compatible with the equivalent command introduced in Redis 8.2.0.

The command supports three deletion modes:

- `KEEPREF` (default): Deletes the stream entry but leaves PEL references intact in all consumer groups
- `DELREF`: Deletes the stream entry and forcibly removes it from all consumer group PELs
- `ACKED`: Only deletes the entry once every consumer group has acknowledged or passed it — safe for fan-out topologies

The command returns a per-ID integer array: `1` for deleted, `2` for exists-but-not-yet-deletable (ACKED mode only), and `-1` for not found.

I've included test coverage for all modes, multi-group scenarios, replication behavior, syntax, and edge cases.

This is my first PR (alongside `XACKDEL` #3466). And I've tried to follow the style guide & contributing guidelines. Happy to make any corrections if I've missed something.

It's possible to create a shared function for parsing the mode & id count between `XACKDEL` and `XDELEX`. So I can follow up with a PR to reduce that duplicated code. Or, I can merge these 2 PRs into one.

### CODE REVIEW

**Comment by nickiaq** [comment]
The force push a few seconds ago was just keeping this up with unstable.

File: src/t_stream.c [hunk 1]
```diff
@@ -3558,6 +3558,193 @@ void xdelCommand(client *c) {
...
+    }
+
+    /* If missing stream, return -1 for each ID. */
+    if (o == NULL) {
+        addReplyArrayLen(c, id_count);
+        for (int i = 0; i < id_count; i++) {
+            addReplyLongLong(c, -1);
+        }
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/t_stream.c:3640]: Fixed in b9af80fed.

File: src/t_stream.c [hunk 2]
```diff
@@ -3558,6 +3558,193 @@ void xdelCommand(client *c) {
...
+            streamGetEdgeID(s, 1, 1, &s->first_id);
+        }
+    }
+
+    /* Propagate the write if needed. */
+    if (deleted) {
+        signalModifiedKey(c, c->db, c->argv[1]);
+        notifyKeyspaceEvent(NOTIFY_STREAM, "xdel", c->argv[1], c->db->id);
+        server.dirty += deleted;
+    }
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/t_stream.c:3754]: b9af80fed implements sending keyspace events in the case of removing orphaned PEL entries. Although there is the inconsistent behavior around this as noted in issue #3429, it seems like the correct behavior is to notify on PEL ref only change. So that's what's implemented to avoid adding another case to the list to fix from #3429.

**Comment by hpatro** [comment]
@valkey-review-bot Please review this change.

File: src/commands/xdelex.json [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                "type": "pure-token"
+            },
+            {
+                "name": "numids",
+                "type": "integer"
+            },
+            {
+                "name": "id",
+                "token": "IDS",
+                "type": "string",
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/commands/xdelex.json:78]: Whoops! I fixed this in `XACKDEL` but not here. I’m still learning the command JSON DSL and was having trouble getting the optional key setup here. From what I can tell, the `token` on `MODE` is the issue here. Removing that, we get the behavior that’s needed here for an optional `MODE` parameter with no `MODE` token itself, just `XDELEX <key> [KEEPREF | DELREF | ACKED] …`.

Fixed in commit: 703506c

File: src/t_stream.c [hunk 4]
```diff
@@ -3558,6 +3558,201 @@ void xdelCommand(client *c) {
...
+            }
+            first_loop = 0;
+        }
+        raxStop(&ri_cgroups);
+    }
+
+    /* Respond with array of results and perform stream deletions. */
+    int deleted = 0;
+    bool first_entry = 0;
+    addReplyArrayLen(c, id_count);
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/t_stream.c]: I've addressed this in 703506c as the bot recommends: by moving the `addReplyArrayLen` and `addReplyLongLong` to a separate loop after marking `server.dirty` and emitting signals.

**Comment by nickiaq** [comment]
Thank you for starting the review for this as well as XACKDEL! These were great comments from the bot. Added comments in thread on each. And, I'll add responses on the XACKDEL one hopefully shortly as well.

File: src/t_stream.c [hunk 5]
```diff
@@ -3558,6 +3558,212 @@ void xdelCommand(client *c) {
...
+    } else if (strcasecmp(objectGetVal(c->argv[argi]), "DELREF") == 0) {
+        argi += 1;
+        mode = 1;
+    } else if (strcasecmp(objectGetVal(c->argv[argi]), "ACKED") == 0) {
+        argi += 1;
+        mode = 2;
+    } else if (strcasecmp(objectGetVal(c->argv[argi]), "IDS") != 0) {
+        addReplyError(c, "xdelex mode not recognized");
+        return;
+    }
```
Comment (zuiderkwast) [inline src/t_stream.c]: Same comments as for the XACKDEL PR.

```suggestion
    }
```

File: src/t_stream.c [hunk 6]
```diff
@@ -3558,6 +3558,212 @@ void xdelCommand(client *c) {
...
+
+    /* Expect IDS token. */
+    if (strcasecmp(objectGetVal(c->argv[argi]), "IDS") != 0) {
+        addReplyError(c, "syntax error");
```
Comment (zuiderkwast) [inline src/t_stream.c]:
```suggestion
        addReplyErrorObject(c, shared.syntaxerr);
```

File: src/t_stream.c [hunk 7]
```diff
@@ -3558,6 +3558,212 @@ void xdelCommand(client *c) {
...
+    }
+    argi++; /* past numids */
+
+    /* Validate numids matches remaining arg count. */
+    long long actual_ids = c->argc - argi;
+    if (id_count > actual_ids) {
+        addReplyError(c, "numids parameter must match the number of IDs provided");
+        return;
+    } else if (id_count < actual_ids) {
+        addReplyError(c, "syntax error");
```
Comment (zuiderkwast) [inline src/t_stream.c]:
```suggestion
    if (id_count != actual_ids) {
        addReplyErrorObject(c, shared.syntaxerr);
```

**Review by zuiderkwast — COMMENTED** [review]
Looks good in general and very similar to the XACKDEL PR. The two commands looked identical to me, but finally I noticed that the `group` argument is absent in XDELEX.

Since it has much in common with the other PR and we prefer to avoid duplicated logic, consider putting both commands in one PR.

@coderabbitai Please check for duplicate code with XDEL and with the PR https://github.com/valkey-io/valkey/pull/3466 which adds XACKDEL. How can the code be de-duplicated, for example using a helper function?

File: src/t_stream.c (hunk 7 above)
Comment (nickiaq) [reply src/t_stream.c]: Fixed in 8e5e202; addressed as recommended here, but also needed to change tests to match.

File: src/t_stream.c (hunk 6 above)
Comment (nickiaq) [reply src/t_stream.c]: Fixed in 8e5e202!

File: src/t_stream.c (hunk 5 above)
Comment (nickiaq) [reply src/t_stream.c]: Indeed, thanks for simplifying. Fixed in 8e5e202.

**Comment by nickiaq** [comment]
Same applies here as in [this comment](https://github.com/valkey-io/valkey/pull/3466#issuecomment-5594200126); added `"since":"9.2.0"` and setup a new PR with this plus #3466 and a new commit combining common code across all 3 commands into shared helpers.

**Comment by nickiaq** [comment]
Closing this in favor of #4629!
